File: Fundamental courses of the major/Math For Computer Science/TAI LIEU KHOA TRUOC/C++ & python/C4 Optimization/Gradient Descent 1 var.py
from __future__ import division, print_function, unicode_literals
import math
import numpy as np
import matplotlib.pyplot as plt
def grad(x):
    return 2*x+ 5*np.cos(x)

def cost(x):
    return x**2 + 5*np.sin(x)

def myGD1(alpha, x0, gra = 1e-3, loop = 1000):
    x = [x0]
    for it in range(loop):
        x_new = x[-1] - alpha*grad(x[-1])
        if abs(grad(x_new)) < gra:
            break
        x.append(x_new)
    return (x, it)

if __name__ == '__main__':
    (x1, it1) = myGD1(.1, -10)
    (x2, it2) = myGD1(.1, 10)
    print('Solution x1 = %f, cost = %f, obtained after %d iterations'%(x1[-1], cost(x1[-1]), it1))
    print('Solution x2 = %f, cost = %f, obtained after %d iterations'%(x2[-1], cost(x2[-1]), it2))

    (x3, it3) = myGD1(.5, 10)
    print('Solution x3 = %f, cost = %f, obtained after %d iterations'%(x3[-1], cost(x3[-1]), it3))

    (x4, it4) = myGD1(.01, 10)
    print('Solution x4 = %f, cost = %f, obtained after %d iterations'%(x4[-1], cost(x4[-1]), it4))

File: Fundamental courses of the major/Math For Computer Science/TAI LIEU KHOA TRUOC/C++ & python/C4 Optimization/Gradient Descent 1 var.cpp
#include <iostream>
#include <cmath>
#include <vector>
using namespace std;

double grad(double x) {
    return 2 * x + 5 * cos(x);
}

double cost(double x) {
    return x * x + 5 * sin(x);
}

pair<vector<double>, int> myGD1(double alpha, double x0, double gra = 1e-3, int loop = 1000) {
    vector<double> x = {x0};
    for (int it = 0; it < loop; ++it) {
        double x_new = x.back() - alpha * grad(x.back());
        if (abs(grad(x_new)) < gra) {
            return make_pair(x, it);
        }
        x.push_back(x_new);
    }
    return make_pair(x, loop);
}

int main() {
    auto [x1, it1] = myGD1(0.1, -10);
    auto [x2, it2] = myGD1(0.1, 10);
    cout << "Solution x1 = " << x1.back() << ", cost = " << cost(x1.back()) << ", obtained after " << it1 << " iterations" << endl;
    cout << "Solution x2 = " << x2.back() << ", cost = " << cost(x2.back()) << ", obtained after " << it2 << " iterations" << endl;

    auto [x3, it3] = myGD1(0.5, 10);
    cout << "Solution x3 = " << x3.back() << ", cost = " << cost(x3.back()) << ", obtained after " << it3 << " iterations" << endl;

    auto [x4, it4] = myGD1(0.01, 10);
    cout << "Solution x4 = " << x4.back() << ", cost = " << cost(x4.back()) << ", obtained after " << it4 << " iterations" << endl;

    return 0;
}
